# Método de Runge-Kutta de 4to orden

Clase: F1014B Modelación Computacional de Sistemas Electromagnéticos

Autor: Edoardo Bucheli

Profesor de Cátedra, Tec de Monterrey Campus Santa Fe

## Introducción
En esta sesión aprenderemos un método numérico para la solución de problemas de valor inicial con la siguiente forma,

$$\frac{dy}{dx} = f(x,y)\qquad y(x_0) = y_0$$

El método que estudiaremos se conoce como el método de Runge-Kutta desarrollado por los matemáticos alemanes Carl Runge y Wilhem Kutta.

Este método es a su vez una versión más precisa del **Método de Euler** el cual estudiaremos rápidamente antes de pasar a Runge-Kutta.

## Método de Euler

El método de Euler se basa en un principio muy sencillo. Dado un campo de pendientes, creamos una trayectoria que se mueve conforme a la pendiente en un punto $(x,y)$ dado un movimiento horizontal $\Delta x = h$. Este procedimiento se visualiza en la siguiente imágen.

<img src="imgs/euler_method.png">


En la imágen se aprecia de manera muy clara el error que ocurre al llevar a cabo la aproximación. Para mejorar la aproximación, podemos reducir el tamaño de $h$ como se muestra en la siguiente figura,

<img src="imgs/euler_2.png">

En la imagen anterior se muestran aproximaciones con $h = 1$, $h = 0.2$ y $h = 0.05$.

Aunque un valor para $h$ puede ser lo suficientemente chico para obtener un resultado, si el valor final de $x$ que estamos buscando es muy grande entonces nuestro algoritmo puede ser muy costoso computacionalmente. Es por ello que resulta necesario encontrar un método más eficiente.

## Ejercicio: Implementa el método de Euler

Ahora si empezaremos a programar, en esta sección necesitas implementar el Método de Euler.

Para hacer eso definamos el método formalmente:


### Definición: Método de Euler

Para el problema de valor inicial,

$$\frac{dy}{dx} = f(x,y)\qquad y(x_0)=y_0$$

El método de Euler con salto de tamaño $h$ consiste en aplicar la fórmula iterativa,

$$y_{n+1} = y_n + h\cdot f(x_n,y_n)$$
$$x_{n+1} = x_n + h$$

para calcular las aproximaciones $y_1,y_2,y_3,\dots$ de los valores reales $y(x_1),y(x_2),y(x_3),\dots$ que emanan de la solución exacta $y(x)$

### Solución Guiada

Empecemos por importar las librerías necesarias.

In [9]:
import numpy as np
import matplotlib.pyplot as plt

Para la solución del problema, intentemos hacer una implementación similar a lo que haría un/a programador/a con más experiencia.

Por lo tanto separaremos la solución en dos funciones. Una se encargará de calcular un paso del método de euler y la segunda utilizará esta función para repetir el proceso tantas veces sea necesario.

Empecemos entonces por implementar la función `euler_step()` que se encargará de llevar a cabo un paso del método de Euler.

In [10]:
def euler_step(x_n,y_n,f ,h):
    """
    Calcula un paso del método de euler
    Entradas:
    x_n: int,float
        El valor inicial de x en este paso
    y_n: int, float
        El valor inicial de y en este paso
    f: función
        Una función que represente f(x,y) en el problema de valor inicial.
    h: int, float
        El tamaño del salto de un paso al siguiente
        
    Salida:
    x_n_plus_1: float
        El valor de x actualizado de acuerdo al método de euler
    y_n_plus_1: float
        El valor de y actualizado de acuerdo al método de euler
    """
    
    # Empieza tu código aquí (alrededor de 2 líneas)
    x_n_plus_1 = x_n + h
    y_n_plus_1 = f(x_n + y_n, y_n)
    
    return x_n_plus_1, y_n_plus_1
    

Probemos tu función con el siguiente problema,

$$\frac{dy}{dx} = x + \frac{1}{5}y \qquad y(0) = -3 $$

Utilizando $h = 1$

In [11]:
x_0 = 0
y_0 = -3

def f(x,y):
    return x + (1/5)*y

h = 1

print(euler_step(x_0,y_0,f,h))

(1, -3.6)


Tu resultado debería ser `(1,-3.6)`

Ahora que tenemos una función que calcula un paso del método, implementemos la función `euler_method()` que use la función `euler_step()` para generar una lista de valores hasta una nueva variable `x_goal`

In [12]:
def euler_method(x_0,y_0,x_goal,f,h):
    """
    Regresa una lista para aproximaciones de y con el método de euler hasta un cierto valor x_goal
    Entradas:
    x_n: int,float
        El valor inicial de x
    y_n: int, float
        El valor inicial de y
    x_goal: int,float
        El valor hasta donde queremos calcular las aproximaciones del método de euler
    f: función
        Una función que represente f(x,y) en el problema de valor inicial.
    h: int, float
        El tamaño del salto de un paso al siguiente
        
    Salida:
    x_n_list: list
        Una lista con los valores de 'x' desde x_0 hasta el valor x_{n+1} más cercano a x_goal que sea también menor
    y_n_list: list
        Una lista con las aproximaciones 'y' evaluadas con los valores de x desde x_0 hasta el valor x_n+1 más cercano a x_goal que sea también menor
    """
    
    # Definimos x_n y y_n como los valores iniciales
    x_n = x_0
    y_n = y_0
    
    # Crea las listas x_n_list y y_n_list por ahora solo contienen los valores iniciales
    x_n_list = [x_0]
    y_n_list = [y_0]
    
    # Crea aquí un ciclo donde lleves a cabo el procedimiento tantas veces sea necesario,
    # no olvides guardar cada valor que calcules en las listas 'x_n_list' y 'y_n_list'
    # Aprox 4 líneas
    for i in range(x_n, x_goal, h):
        x,y = euler_step(x_n_list[-1], y_n_list[-1], f, h )
        x_n_list.append(x)
        y_n_list.append(y)
        
    return x_n_list,y_n_list

In [13]:
x_0 = 0
y_0 = -3

def f(x,y):
    return x + (1/5)*y

h = 1
x_goal = 5

print(euler_method(x_0,y_0,x_goal,f,h))

([0, 1, 2, 3, 4, 5], [-3, -3.6, -3.3200000000000003, -1.9840000000000004, 0.6191999999999995, 4.743039999999999])


La salida de la celda anterior debería ser:

`([0, 1, 2, 3, 4, 5], [-3, -3.6, -3.3200000000000003, -1.9840000000000004, 0.6191999999999993, 4.743039999999999])`

## Un Método de Euler Mejorado

Una manera sencilla de mejorar el método de Euler será obteniendo más de una pendiente y tomando el promedio de las pendientes para mejorar la predicción como se muestra en la siguiente figura.

<img src="imgs/euler_3.png">

Vemos que una vez que generamos un nuevo punto $y(x_{n+1})$ podemos obtener la pendiente en este punto, tomar el promedio entre esta pendiente y la encontrada en $y(x_n)$ para encontrar una nueva predicción un poco mejor.

## Método de Runge-Kutta

Utilizando el **Teorema Fundamental del Cálculo** podemos derivar la siguiente expresión,

$$y(x_{n+1})-y(x_n) = \int_{x_n}^{x_{n+1}}y'(x)dx$$

Y a su vez, por la **Ley de Simpson** de integración podemos aproximar esto como,

$$y(x_{n+1})-y(x_n)\approx \frac{h}{6}\bigg[y'\big(x_n\big)+4y'\bigg(x_n+\frac{h}{2}\bigg)+y'\big(x_{n+1}\big)\bigg]$$

Y por lo tanto podemos despejar $y_{n+1}$ de la siguiente forma,

$$y(x_{n+1})\approx y(x_n) + \frac{h}{6}\bigg[y'\big(x_n\big)+2y'\bigg(x_n+\frac{h}{2}\bigg)+2y'\bigg(x_n+\frac{h}{2}\bigg)+y'\big(x_{n+1}\big)\bigg]$$

Definiremos los términos dentro del paréntesis $y'\big(x_n\big)$, $y'\big(x_n+\frac{h}{2}\big)$, $y'\big(x_n+\frac{h}{2}\big)$ y $y'\big(x_{n+1}\big)$ como $k_1,k_2,k_3,k_4$ respectivamente de la siguiente forma,

### $k_1 = f(x_n,y_n)$

Esto es la pendiente en $(x_n,y_n)$ misma pendiente que utilizaríamos en el Método de Euler original.

### $k_2 = f(x_n+\frac{1}{2}h,y_n+\frac{1}{2}hk_1)$

Esto es la pendiente en el punto medio de $x_n$ y $x_{n+1}$ de acuerdo a la pendiente $k_1$
 
### $k_3 = f(x_n+\frac{1}{2}h,y_n+\frac{1}{2}hk_2)$

Esto es una correción de la pendiente del punto medio de $x_n$ y $x_{n+1}$ como se hace en el método de Euler corregido.

### $k_4 = f(x_{n+1},y_n+hk_3)$

Esto es la pendiente del punto $(x_{n+1},y_{n+1})$ basado en la pendiente corregida $k_3$.

Lo que finalmente nos lleva a la forma,

$$y_{n+1}=y_n+\frac{h}{6}(k_1+2k_2+2k_3+k_4)$$

Nota que se usan las cuatro pendientes calculadas de manera ponderada. Es decir que no es exactamente un promedio, sino que le damos un poco más de peso a ciertas pendientes. Específicamente a las de los puntos medios.

Definamos entonces de manera formal el algoritmo.

### Definición: Método de Runge-Kutta

Para el problema de valor inicial,

$$\frac{dy}{dx} = f(x,y)\qquad y(x_0)=y_0$$

El método de Runge-Kutta con salto de tamaño $h$ consiste en aplicar la fórmula iterativa,

$$y_{n+1}=y_n+\frac{h}{6}(k_1+2k_2+2k_3+k_4)$$

Dónde,

* $k_1 = f(x_n,y_n)$
* $k_2 = f(x_n+\frac{1}{2}h,y_n+\frac{1}{2}hk_1)$
* $k_3 = f(x_n+\frac{1}{2}h,y_n+\frac{1}{2}hk_2)$
* $k_4 = f(x_{n+1},y_n+hk_3)$

para calcular las aproximaciones $y_1,y_2,y_3,\dots$ de los valores reales $y(x_1),y(x_2),y(x_3),\dots$ que emanan de la solución exacta $y(x)$

## Ejercicio: Implementa el Método de Runge-Kutta

Ahora implementaremos el método de Runge-Kutta. Al igual que Euler convendrá hacerlo como una función que podamos llamar para implementar un paso combinada de una función que implemente un cierto número de iteraciones.

Empecemos por definir la función `runge_kutta_step()` que llevará a cabo una iteración del método.

In [14]:
def runge_kutta_step(x_n,y_n,f,h):
    """
    Calcula una iteración del método de Runge-Kutta de cuarto orden
    Entradas:
    x_n: int,float
        El valor inicial de x en este paso
    y_n: int, float
        El valor inicial de y en este paso
    f: función
        Una función que represente f(x,y) en el problema de valor inicial.
    h: int, float
        El tamaño del salto de un paso al siguiente
        
    Salida:
    x_n_plus_1: float
        El valor de x actualizado de acuerdo al método de euler
    y_n_plus_1: float
        El valor de y actualizado de acuerdo al método de euler
    """
    x_n_plus_1 = x_n + h

    # Calcula cada una de las pendientes k de acuerdo al método de Runge-Kutta
    k_1 = f(x_n, y_n) #
    k_2 = f(x_n + h/2, y_n + (h/2 * k_1)) #
    k_3 = f(x_n + h/2, y_n + (h/2 * k_2)) #
    k_4 = f(x_n_plus_1 , y_n + h*k_3) #
    
    # Y ahora calcula y_n_plus_1 y x_n_plus_1
    # Aprox 2 líneas
    y_n_plus_1 = y_n + (h/6)*(k_1 + 2*k_2 + 2*k_3 + k_4)
    
    return x_n_plus_1, y_n_plus_1

In [15]:
def f(x,y):
    return x + y

x_0 = 0
y_0 = 1
h = 0.5

In [16]:
x_1_test, y_1_test = runge_kutta_step(x_0,y_0,f,h)
print(f"x_1 = {x_1_test}\ny_1 = {y_1_test}")

x_1 = 0.5
y_1 = 1.796875


El resultado anterior debería ser:
```
x_1 = 0.5
y_1 = 1.796875
```

Para terminar, implementemos la función `runge_kutta()`

In [17]:
def runge_kutta(x_0,y_0,x_goal,f,h):
    """
    Regresa una lista para aproximaciones de y con el método de Runge-Kutta hasta un cierto valor x_goal
    Entradas:
    x_n: int,float
        El valor inicial de x
    y_n: int, float
        El valor inicial de y
    x_goal: int,float
        El valor hasta donde queremos calcular las aproximaciones del método de euler
    f: función
        Una función que represente f(x,y) en el problema de valor inicial.
    h: int, float
        El tamaño del salto de un paso al siguiente
        
    Salida:
    x_n_list: list
        Una lista con los valores de 'x' desde x_0 hasta el valor x_{n+1} más cercano a x_goal que sea también menor
    y_n_list: list
        Una lista con las aproximaciones 'y' evaluadas con los valores de x desde x_0 hasta el valor x_n+1 más cercano a x_goal que sea también menor
    """
    
    # Definimos x_n y y_n como los valores iniciales
    x_n = x_0
    y_n = y_0
    
    # Crea las listas x_n_list y y_n_list por ahora solo contienen los valores iniciales
    x_n_list = [x_0]
    y_n_list = [y_0]
    
    # Crea aquí un ciclo donde lleves a cabo el procedimiento tantas veces sea necesario,
    # no olvides guardar cada valor que calcules en las listas 'x_n_list' y 'y_n_list'
    # Aprox 4 líneas
    
    # for i in range(x_n, x_goal, h):
    while x_n<=x_goal:

        x,y = runge_kutta_step(x_n_list[-1], y_n_list[-1], f, h )
        x_n_list.append(x)
        y_n_list.append(y)
        x_n+=h
        
    return x_n_list,y_n_list

In [19]:
def f(x,y):
    return x + y

x_0 = 0
y_0 = 1
h = 0.1
x_goal = 1


x_list, y_list = runge_kutta(x_0,y_0,x_goal,f,h)

Usemos una librería llamada `prettytable` para imprimir nuestro resultado. Si no la tienes instalada entonces la siguiente linea arrojaría un error.

In [120]:
#!pip install PrettyTable

In [21]:
from prettytable import PrettyTable

En caso de que no tengas la librería la puedes instalar corriendo el siguiente comando en una celda,

`!pip install PrettyTable`

O puedes simplemente correr el comando `pip install PrettyTable` en una ventana de la terminal (mac y linux) o el prompt de anaconda (windows).

In [22]:
mytable = PrettyTable()

for x,y in zip(x_list,y_list):
    mytable.add_row(["{:0.2f}".format(x),"{:0.6f}".format(y)])
print(mytable)

+---------+----------+
| Field 1 | Field 2  |
+---------+----------+
|   0.00  | 1.000000 |
|   0.10  | 1.110342 |
|   0.20  | 1.242805 |
|   0.30  | 1.399717 |
|   0.40  | 1.583648 |
|   0.50  | 1.797441 |
|   0.60  | 2.044236 |
|   0.70  | 2.327503 |
|   0.80  | 2.651079 |
|   0.90  | 3.019203 |
|   1.00  | 3.436559 |
|   1.10  | 3.908327 |
+---------+----------+


La salida de la celda anterior debería ser:
```
+---------+----------+
| Field 1 | Field 2  |
+---------+----------+
|   0.00  | 1.000000 |
|   0.10  | 1.110342 |
|   0.20  | 1.242805 |
|   0.30  | 1.399717 |
|   0.40  | 1.583648 |
|   0.50  | 1.797441 |
|   0.60  | 2.044236 |
|   0.70  | 2.327503 |
|   0.80  | 2.651079 |
|   0.90  | 3.019203 |
|   1.00  | 3.436559 |
|   1.10  | 3.908327 |
+---------+----------+
```